File: 2020/11/week4/minhyungPark/algorithm/LeetCode1657.kt
class LeetCode1657 {

    fun closeStrings(word1: String, word2: String): Boolean {
        if (word1.length != word2.length) {
            return false
        }

        val arr1 = IntArray(26)
        val arr2 = IntArray(26)
        val set1 = mutableSetOf<Char>()
        val set2 = mutableSetOf<Char>()

        for (i in word1.indices) {
            arr1[word1[i] - 'a']++
            arr2[word2[i] - 'a']++
        }

        word1.toCollection(set1)
        word2.toCollection(set2)


        val list1 = arr1.filter { it != 0 }.toList().sorted()
        val list2 = arr2.filter { it != 0 }.toList().sorted()

        return list1 == list2 && set1 == set2
    }

}

File: 2020/11/week5/1ilsang/algorithm/leetcode1657.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {boolean}
 */
const closeStrings = (word1, word2) => {
  if (word1.length != word2.length) return false;
  const arr1 = Array(26).fill(0);
  const arr2 = Array(26).fill(0);

  for (let i = 0; i < word1.length; i++) {
    arr1[word1[i].charCodeAt() - 97]++;
    arr2[word2[i].charCodeAt() - 97]++;
  }

  for (let i = 0; i < 26; i++) {
    if ((arr1[i] && !arr2[i]) || (!arr1[i] && arr2[i])) return false;
  }
  arr1.sort((a, b) => b - a);
  arr2.sort((a, b) => b - a);

  for (let i = 0; i < 26; i++) {
    if (arr1[i] != arr2[i]) return false;
  }
  return true;
};
